can-j1939-install-kernel-module.md

```diff
@@ -0,0 +1,162 @@
+# can-j1939 kernel module installation #
+
+
+
+### Problem
+
+You already have **can0** or **vcan0** up and working, **can-utils** downloaded and compiled to **~/can/can-utils** and you can send and receive frames without problems. However, when you want to bring up **can-j1939** you get error like this:
+
+```bash
+avra@vm-debian:~/can/can-utils$ sudo modprobe can-j1939
+modprobe: FATAL: Module can-j1939 not found in directory /lib/modules/5.7.0.0.bpo.2-amd64
+```
+
+and also this:
+
+```bash
+avra@vm-debian:~/can/can-utils$ testj1939
+testj1939: socket(j1939): Protocol not supported
+```
+
+
+
+### Solution
+
+Above errors mean that **can-j1939** was not enabled in your kernel and you need to do it manually. There are several ways to do it. Any Linux kernel since 5.4 has **can-j1939** module, but you will probably want to install fresher version, which leads to downloading kernel sources, enabling **can-j1939** module, recompiling kernel and installing it. I will be using Debian 10.5 x64 (buster testing) virtual machine.
+
+
+
+#### 1. Download kernel source ####
+
+We will download Debian patched kernel 5.8. First update your sources
+
+```avra@vm-debian:~$ sudo apt update```
+
+and then see if what Debian pathed kernel source versions are available
+
+```
+avra@vm-debian:~$ apt-cache search linux-source
+linux-source-4.19 - Linux kernel source for version 4.19 with Debian patches
+linux-source - Linux kernel source (meta-package)
+linux-source-5.4 - Linux kernel source for version 5.4 with Debian patches
+linux-source-5.5 - Linux kernel source for version 5.5 with Debian patches
+linux-source-5.6 - Linux kernel source for version 5.6 with Debian patches
+linux-source-5.7 - Linux kernel source for version 5.7 with Debian patches
+linux-source-5.8 - Linux kernel source for version 5.8 with Debian patches
+```
+
+If kernel 5.8 does not show in the list (I can see it since I have already upgraded my 4.19 kernel to backported 5.7), then you will need to add backports to your sources list. Here is how my **/etc/apt/sources.list** looks like (you will need to append at least last line to yours):
+

```

Comment:
I am not sure that stock Debian will find all needed packages, so I wanted to show the user how my working sources.list file looks like. 
If you think it's too much, then maybe something like
```sudo echo 'deb http://deb.debian.org/debian buster-backports main contrib non-free' >> /etc/apt/sources.list```
would be better and simpler?

Reply:
> I am not sure that stock Debian will find all needed packages, so I wanted to show the user how my working sources.list file looks like.

Right - with modern Debian based systems, there's no need  to modify the standard `sources.list`. Better create a new one,

```
sudo echo 'deb http://deb.debian.org/debian buster-backports main contrib non-free' >> /etc/apt/sources.list
```

This does not work, as the `sudo` will act on the `echo` not on the `>>`, use this trick instead:

```
 echo 'deb http://deb.debian.org/debian buster-backports main contrib non-free' | sudo tee -a /etc/apt/sources.list
```

`tee -a` for append. But as said above, no need to modify the original `sources.list`, better create a new one:

```
echo 'deb http://deb.debian.org/debian buster-backports main contrib non-free' | sudo tee /etc/apt/sources.list.d/debian-backports.list
```

Reply:
I have just tested and 'sudo echo...' gives permission denied error and only works after file permissions are extended so I will not make it that way to avoid confusion.

Reply:
See above `sudo echo` does not work.

Instead of `sudo echo "foo" > bar` use `echo foo | sudo tee bar`
Instead of `sudo echo "baz" >> qux` use `echo baz | sudo tee -a qux`

Reply:
Do you know if I need to put **sudo reboot** after adding backports? I am asking because **apt-cache linux-image** now report duplicate sources although I have deleted that line from **sources.list**.

Reply:
Reboot doesn't do any change to the `sources.list` etc. You have to re-download the packages information with `apt update`.

Reply:
It seams to work well now :-)